### DISCUSSION TITLE
How to use LightningCLI to resume training from checkpoint

### DISCUSSION THREAD

**Discussion by KirillShmilovich** [Lightning-AI/pytorch-lightning#16845] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I'm trying to use the LightningCLI to resume training from checkpoint, but don't understand how to pass keyword arguments of the checkpoint path to enable restoring the model and trainer states from the checkpoint, my current workaround is the following:

```python
def cli_main():
    class MyCLI(LightningCLI):
        def add_arguments_to_parser(self, parser):
            parser.link_arguments("data.unmask_index", "model.unmask_index", apply_on="instantiate")

    ckpt_path = 'my/checkpoint/path/ckpt.ckpt'
    cli = MyCLI(
        DDPM,
        InMemoryDataModule,
        save_config_overwrite=True,
        seed_everything_default=4321,
        run=False,  # used to de-activate automatic fitting.
        trainer_defaults={
            "accelerator": "gpu",
            "devices": 1,
            "max_epochs": 2500000,
            "enable_progress_bar": False,
            #"limit_val_batches": 0.25,
            "check_val_every_n_epoch": 30,
            # "num_sanity_val_steps": 0,
            # "logger": TensorBoardLogger(save_dir="./",name="DDPM_timesteps"),
        },
    )
    cli.trainer.fit(cli.model.load_from_checkpoint(ckpt_path), datamodule=cli.datamodule, ckpt_path=ckpt_path)
```

note that I call `cli.trainer.fit(cli.model.load_from_checkpoint(ckpt_path), datamodule=cli.datamodule, ckpt_path=ckpt_path)` specifying the checkpoint path. 

***Is there a better/recommended method for dealing with restarting from checkpoints with the LightningCLI?***

**Comment by tshu-w** [comment]
Just `python cli.py fit --ckpt_path PATH`, you can `python cli.py fit --help` to check all available options.

**Reply by adamjstewart** [reply]
Is there a way to automatically select the most recent checkpoint or best checkpoint? I would like to do this but don't always know the filepath to my checkpoint file.

**Reply by adamjstewart** [reply]
`--ckpt_path last` doesn't seem to work for me, although it's in the help message.

**Comment by KirillShmilovich** [comment]
Then only option involving a checkpoint appears to be in the `trainer.resume_from_checkpoint` option that appears to now be deprecated -- this also appears to not address the issue of reinstating the model state & weights from the checkpoint, because this option presumably only impacts the trainer state?  

![image](https://user-images.githubusercontent.com/40403472/220829918-a8cc30b9-09f2-43d6-a2cd-be15ca58ff81.png)

**Reply by tshu-w** [reply]
The `--ckpt_path` option is at the end of `--help` message, what's your PL version?
and please 
<img width="1886" alt="Screenshot 2023-02-23 at 18 35 17" src="https://user-images.githubusercontent.com/13161779/220883042-a24ac698-cfb4-4894-899c-4aa9b70d1b01.png">

Oh, I see, `run=False` doesn't support it, I'm not sure why.

**Reply by KirillShmilovich** [reply]
I see, I now see this `ckpt_path` option is available when removing `run=False`. Do you know if removing the `run=False` and running `python cli.py fit ...` is equivalent to my current setup with the manual `cli.trainer.fit`?

**Reply by mauvilsa** [reply]
With `run=True` the parser has subcommands, one for each of the trainer methods. With `run=False` there is no `ckpt_path` option because that is a parameter of the `fit` method. You can keep `run=False` but you have to implement the `ckpt_path` functionality yourself. In `add_arguments_to_parser` add an argument and then use the value for this argument when running `fit`.

**Reply by KirillShmilovich** [reply]
But my question is if specifying the `ckpt_path` (with `run=True`) will also instantiate the model state & weights alongside the trainer state?

**Reply by mauvilsa** [reply]
I would say yes, considering the description of [`ckpt_path`: "Path/URL of the checkpoint from which training is resumed"](https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.trainer.trainer.Trainer.html#pytorch_lightning.trainer.trainer.Trainer.fit).

If you run `LightningCLI` with `run=True` and `fit --ckpt_path=path` internally what is run is `cli.trainer.fit(model=cli.model, datamodule=cli.datamodule, ckpt_path=path)`. I would guess that the `.load_from_checkpoint(ckpt_path)` in the code of the question is unnecessary.

**Reply by KirillShmilovich** [reply]
This worked! Thanks!

**Reply by adosar** [reply]
@mauvilsa If we call `fit --config=config.yaml --ckpt_path=path` and `config.yaml` configures optimizer, does this interferes in any way with the state of the optimizer saved in `path` (due to checkpointing)? Or `path` just load the LightningModule’s `state_dict`? This might be the case if one wants to resume training but with a different optimizer.

**Reply by mauvilsa** [reply]
@adosar I am only aware about how the CLI works, and it just provides parameters and instantiates classes. What you want to know depends on what `fit` does internally when it receives a `ckpt_path` argument. I am not sure if this is clearly explained in the docs or you will have to look at the code or experiment to understand what exactly happens.

**Reply by adosar** [reply]
@mauvilsa Yeap, I would need to look how the trainer works under the hood.

**Reply by senarvi** [reply]
[This document](https://lightning.ai/docs/pytorch/stable/common/checkpointing_basic.html) explains that `resume_from_checkpoint` has been deprecated in Lightning >= 1.0.0 and `ckpt_path`should be used to resume training from a checkpoint. This will restore the full training, i.e. model weights, epoch, step, LR schedulers, etc.

**Comment by KasuganoLove** [comment]
Hello! There is a similar question. I have already trained an encoder saved in encoder_ckpt.ckpt. I want to train another autoencoder using encoder_ckpt.ckpt as the frozen encoder. How should I configure the LightningCLI?
